bsip-0035.md

```diff
@@ -0,0 +1,306 @@
+$19 / 50 CORE = $0.38, so based on the price, Alice and Bob should match.
+
+Bob's $10 / $0.38 ~ 26.3. So 26.3 is the fewest CORE he is willing to accept
+(assuming that the meaning of "price" is "the least favorable exchange rate
+a party is willing to accept in trade"). Combined with the design restriction
+that satoshis are indivisible, in practice this means Bob will only accept 27
+or more CORE for his $10.
+
+But $10 / 27 gives a price smaller than $0.370 and $0.371, which is smaller
+than Alice's sale price of $0.375. So neither party can fill this offer.
+
+We need to come to a compromise.
+
+## The Possible Solutions
+
+There are some possible solutions listed in the 5th comment of [bitshares-core
+issue #132](https://github.com/bitshares/bitshares-core/issues/132):
+
+- (a) Fill someone at a less favorable exchange rate than the price they
+specified in their order.  Downside:  This violates the above definition of
+price; i.e. if a user enters a price intending the system to never sell below
+that price in any circumstance, the system will not always behave in a way
+which fulfills that user intent.
+
+- (b) Keep both orders on the books.  Downside:  This complicates the matching
+algorithm, as now Alice might be able to match an order behind Bob's order.
+Naive implementation would have potentially unbounded matching complexity;
+a more clever implementation might be possible but would require substantial
+design and testing effort.
+
+- (c) Cancel an order.  This is complicated by the fact that an order such as
+a margin call cannot be cancelled.  Downside:  When there are margin calls
+happening, it seems perverse to delete a large order that's willing to fill
+them just because the lead margin call happens to fall in a narrow window
+which causes a rounding issue.  Also, orders cancelled by this mechanism
+cannot be refunded.  Otherwise an attacker who wants to consume
+a lot of memory on all nodes could create a large number of orders, then
+trigger this case to cancel them all, getting their investment in deferred
+cancellation fees back without paying the cancel op's per-order fee as
+intended.
+
+- (d) Require all orders to use the same denominator.  Altcoin exchanges and
+many real-world markets like the stock market solve this problem by specifying
+one asset as the denominator asset, specifying a "tick" which is the smallest
+unit of price precision, and requiring all prices to conform.
+Downside:  Complicates the implementation of flipped market UI, may require
+re-working part of market GUI, reduces user flexibility, new asset fields
+required to specify precision, if `n` assets exist then `O(n^2)` markets
+could exist and we need to figure out how to determine the precision
+requirement for all of them.
+
+## The Chosen Solution
+
+Current code actually implemented (a) in the first place: when matching two
+orders, if there is a rounding issue, the order with smaller volume will be
```

Comment:
Please specify how order volumes are compared to determine which is "smaller".

Reply:
Done.